### PR TITLE
Refactor constraints

### PR DESCRIPTION
**Pull request by blegat**
This PR

* moves constraints-related code to a constraints.jl file, moving a lot of code out of JuMP.jl;
* merges SingleVariableConstraint, AffExprConstraint and QuadExprConstraint into ScalarConstraint (same for the vector version) removing a lot of duplicated code both from the definition of the constraint and from the build_constraint function calling them.

As a by-product, it improves support for constraints for which the function is a quadratic expression since the code is now shared with affine expression and variable expression.

### CODE REVIEW

File: src/constraints.jl [hunk 1]
```diff
@@ -0,0 +1,207 @@
+#############################################################################
```
Comment (IainNZ) [inline src/constraints.jl]: Can drop this line now, and the standard copyright header.

File: src/constraints.jl [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+# AbstractShape
```
Comment (IainNZ) [inline src/constraints.jl]: I think these lines are redundant with the docstring.

File: src/constraints.jl [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+## Examples
+
+Given a [`SymmetricMatrixShape`](@ref) of vectorized form `[1, 2, 3]`, the
+following code retrieve the matrix `Symmetric(Matrix[1 2; 2 3])`:
+```julia
+reshape([1, 2, 3], SymmetricMatrixShape(2))
+```
+"""
+function reshape end
```
Comment (IainNZ) [inline src/constraints.jl]: Maybe stupid question: Given there is a Base.reshape, does this squash the Base definition in the scope of JuMP from this point on? What does it do to the docstrings?

File: src/constraints.jl [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+Vector for which the vectorized form corresponds exactly to the vector given.
+"""
+struct VectorShape <: AbstractShape end
+reshape(vectorized_form, ::VectorShape) = vectorized_form
+
+##########################################################################
+# Constraint Reference
+# Holds the index of a constraint in a Model.
+# TODO: Rename "m" field (breaks style guidelines).
+struct ConstraintRef{M <: AbstractModel, C, Shape <: AbstractShape}
```
Comment (IainNZ) [inline src/constraints.jl]: Style: are we putting spaces around `<:`?

Comment (blegat) [reply src/constraints.jl]: I thought so as it is a binary operator but we might make an exception for this one

File: src/constraints.jl (hunk 3 above)
Comment (blegat) [reply src/constraints.jl]: It defines a JuMP.reshape which is a different function form `Base.reshape` so `reshape` in the JuMP scope and one need to explicitely write `Base.reshape`

File: src/affexpr.jl [hunk 5]
```diff
@@ -112,6 +112,8 @@ function value(a::GenericAffExpr{T, V}, map::Function) where {T, V}
     ret
 end
 
+constant(aff::GenericAffExpr) = aff.constant
```
Comment (mlubin) [inline src/affexpr.jl]: Is this meant to be a public API or internal only? If it's the former it should have a doc string, if it's the latter then it could use a scarier name like `extract_constant` and comment that it's meant for internal use.

File: src/constraints.jl [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+"""
+    reshape(vectorized_form::Vector, shape::AbstractShape)
+
+Return an object in it original shape `shape` given its vectorized form
+`vectorized_form`.
+
+## Examples
+
+Given a [`SymmetricMatrixShape`](@ref) of vectorized form `[1, 2, 3]`, the
+following code retrieve the matrix `Symmetric(Matrix[1 2; 2 3])`:
```
Comment (mlubin) [inline src/constraints.jl]: Grammar: `retrieves`. Consider `returns` instead though.

File: src/constraints.jl [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+dual_shape(shape::PolynomialShape) = MomentsShape(shape.monomials)
+dual_shape(shape::MomentsShape) = PolynomialShape(shape.monomials)
+```
+"""
+dual_shape(shape::AbstractShape) = shape
+
+"""
+    reshape(vectorized_form::Vector, shape::AbstractShape)
+
+Return an object in it original shape `shape` given its vectorized form
```
Comment (mlubin) [inline src/constraints.jl]: Typo: `in its original shape`.

File: src/constraints.jl [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+function constraint_object(ref::ConstraintRef{Model, MOICON{FuncType, SetType}}) where
+        {FuncType <: MOI.AbstractVectorFunction, SetType <: MOI.AbstractVectorSet}
+    model = ref.m
+    f = MOI.get(model, MOI.ConstraintFunction(), ref)::FuncType
+    s = MOI.get(model, MOI.ConstraintSet(), ref)::SetType
+    return VectorConstraint(jump_function(model, f), s, ref.shape)
+end
+
+# TODO: Find somewhere to put this error message.
+#add_constraint(m::Model, c::Array{AffExprConstraint}) =
```
Comment (mlubin) [inline src/constraints.jl]: Do you know what the intention of this code is? I don't know what error case it handles and why `add_constraint` would say something about macro syntax.

File: src/macros.jl [hunk 9]
```diff
@@ -347,28 +347,21 @@ function parse_constraint(_error::Function, args...)
...
-    return QuadExprConstraint(quad, S(MOIU.getconstant(set)-offset))
+function build_constraint(_error::Function,
+                          expr::Union{GenericAffExpr, GenericQuadExpr},
+                          set::MOI.AbstractScalarSet)
+    offset = constant(expr)
+    add_to_expression!(expr, -offset)
+    return ScalarConstraint(expr, MOIU.shift_constant(set, -offset))
 end
-#build_constraint(x::Vector{<:GenericQuadExpr}, set::MOI.AbstractVectorSet) = VectorQuadExprConstraint(x, set)
+build_constraint(_error::Function, α::Number, set::MOI.AbstractScalarSet) = build_constraint(_error, convert(AffExpr, α), set)
```
Comment (mlubin) [inline src/macros.jl]: Style nit: Split across lines. Same right below.

File: src/sd.jl [hunk 10]
```diff
@@ -41,43 +41,31 @@ function reshape(vectorized_form::Vector{T}, shape::SquareMatrixShape) where T
                         shape.side_dimension)
 end
 
-# Used by the @variable macro. It can also be used with the @constraint macro,
-# this allows to get the constraint reference, e.g.
-# @variable model x[1:2,1:2] Symmetric # x is Symmetric{VariableRef,Matrix{VariableRef}}
-# varpsd = @constraint model x in PSDCone()
+# * If `eltype(Q)` is `AbstractVariableRef`:
```
Comment (mlubin) [inline src/sd.jl]: The formatting of this comment makes it a bit hard to follow. If you're writing pseudo-markdown it may be better to use a docstring instead. Same below.

File: src/constraints.jl (hunk 8 above)
Comment (blegat) [reply src/constraints.jl]: We can remove it, it is done in
https://github.com/JuliaOpt/JuMP.jl/blob/672de56de2e7462dfe454a8b1b716eca40b74f19/src/macros.jl#L362
